File: src/news/news.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Article, ArticleStatus } from './entities/article.entity';
import { NewsCategory } from './entities/news-category.entity';
import { Tag } from './entities/tag.entity';
import { CreateArticleDto, UpdateArticleDto, QueryArticleDto } from './dto/create-article.dto';

@Injectable()
export class NewsService {
  constructor(
    @InjectRepository(Article) private articleRepo: Repository<Article>,
    @InjectRepository(NewsCategory) private newsCatRepo: Repository<NewsCategory>,
    @InjectRepository(Tag) private tagRepo: Repository<Tag>,
  ) {}

  async create(dto: CreateArticleDto, authorId: number) {
    const { tags: tagNames, ...data } = dto;
    const article = this.articleRepo.create({ ...data, authorId });

    if (dto.status === ArticleStatus.PUBLISHED && !dto.publishedAt) {
      article.publishedAt = new Date();
    }

    if (tagNames?.length) {
      article.tags = await this.findOrCreateTags(tagNames);
    }

    return this.articleRepo.save(article);
  }

  async findAll(query: QueryArticleDto) {
    const { page = 1, limit = 10, search, categoryId, tag, status } = query;
    const qb = this.articleRepo.createQueryBuilder('a')
      .leftJoinAndSelect('a.category', 'category')
      .leftJoinAndSelect('a.author', 'author')
      .leftJoinAndSelect('a.tags', 'tags');

    if (search) qb.andWhere('(a.title LIKE :s OR a.excerpt LIKE :s)', { s: `%${search}%` });
    if (categoryId) qb.andWhere('a.categoryId = :categoryId', { categoryId });
    if (tag) qb.andWhere('tags.slug = :tag', { tag });
    if (status) qb.andWhere('a.status = :status', { status });

    qb.addSelect(['author.id', 'author.name', 'author.email'])
      .orderBy('a.createdAt', 'DESC')
      .skip((page - 1) * limit)
      .take(limit);

    const [items, total] = await qb.getManyAndCount();
    return { data: items, meta: { page, limit, total, totalPages: Math.ceil(total / limit) } };
  }

  async findPublic(query: QueryArticleDto) {
    query.status = ArticleStatus.PUBLISHED;
    return this.findAll(query);
  }

  async findOne(id: number) {
    const article = await this.articleRepo.findOne({
      where: { id },
      relations: { category: true, author: true, tags: true },
    });
    if (!article) throw new NotFoundException('Bài viết không tồn tại');
    return article;
  }

  async findBySlug(slug: string) {
    const article = await this.articleRepo.findOne({
      where: { slug, status: ArticleStatus.PUBLISHED },
      relations: { category: true, author: true, tags: true },
    });
    if (!article) throw new NotFoundException('Bài viết không tồn tại');
    return article;
  }

  async update(id: number, dto: UpdateArticleDto) {
    const article = await this.findOne(id);
    const { tags: tagNames, ...data } = dto;

    Object.assign(article, data);

    if (dto.status === ArticleStatus.PUBLISHED && !article.publishedAt) {
      article.publishedAt = new Date();
    }

    if (tagNames !== undefined) {
      article.tags = tagNames.length ? await this.findOrCreateTags(tagNames) : [];
    }

    return this.articleRepo.save(article);
  }

  async remove(id: number) {
    await this.findOne(id);
    await this.articleRepo.delete(id);
    return { success: true, message: 'Xóa bài viết thành công' };
  }

  // News Categories
  async createCategory(data: { name: string; slug: string; sortOrder?: number }) {
    return this.newsCatRepo.save(this.newsCatRepo.create(data));
  }

  async findCategories() {
    return this.newsCatRepo.find({ order: { sortOrder: 'ASC' } });
  }

  async updateCategory(id: number, data: Partial<NewsCategory>) {
    await this.newsCatRepo.update(id, data);
    return this.newsCatRepo.findOneBy({ id });
  }

  async removeCategory(id: number) {
    await this.newsCatRepo.delete(id);
    return { success: true, message: 'Xóa danh mục tin tức thành công' };
  }

  private async findOrCreateTags(names: string[]): Promise<Tag[]> {
    const tags: Tag[] = [];
    for (const name of names) {
      const slug = name.toLowerCase().replace(/[^a-z0-9À-ɏ]+/gi, '-').replace(/^-|-$/g, '');
      let tag = await this.tagRepo.findOne({ where: { slug } });
      if (!tag) {
        tag = await this.tagRepo.save(this.tagRepo.create({ name, slug }));
      }
      tags.push(tag);
    }
    return tags;
  }
}

File: api/src/main/java/com/gese/firesecurity/news/service/NewsService.java
package com.gese.firesecurity.news.service;

import com.gese.firesecurity.common.dto.PageResponse;
import com.gese.firesecurity.news.dto.*;
import com.gese.firesecurity.news.entity.*;
import com.gese.firesecurity.news.repository.ArticleRepository;
import com.gese.firesecurity.news.repository.NewsCategoryRepository;
import com.gese.firesecurity.news.repository.TagRepository;
import jakarta.persistence.EntityNotFoundException;
import jakarta.persistence.criteria.Join;
import jakarta.persistence.criteria.Predicate;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Service
@RequiredArgsConstructor
public class NewsService {

    private final ArticleRepository articleRepo;
    private final NewsCategoryRepository newsCatRepo;
    private final TagRepository tagRepo;

    @Transactional
    public Article create(CreateArticleRequest dto, Long authorId) {
        Article article = Article.builder()
                .title(dto.getTitle()).slug(dto.getSlug())
                .thumbnail(dto.getThumbnail()).content(dto.getContent())
                .excerpt(dto.getExcerpt())
                .status(dto.getStatus() != null ? dto.getStatus() : ArticleStatus.draft)
                .seoTitle(dto.getSeoTitle()).seoDescription(dto.getSeoDescription())
                .authorId(authorId)
                .build();

        if (dto.getCategoryId() != null) article.setCategoryId(dto.getCategoryId());
        if (dto.getStatus() == ArticleStatus.published) {
            article.setPublishedAt(dto.getPublishedAt() != null ? dto.getPublishedAt() : LocalDateTime.now());
        }
        if (dto.getTags() != null && !dto.getTags().isEmpty()) {
            article.setTags(findOrCreateTags(dto.getTags()));
        }

        return articleRepo.save(article);
    }

    public PageResponse<Article> findAll(QueryArticleRequest query) {
        int page = query.getPage() != null ? query.getPage() : 1;
        int limit = query.getLimit() != null ? query.getLimit() : 10;

        Specification<Article> spec = buildSpec(query);
        var pageResult = articleRepo.findAll(spec,
                PageRequest.of(page - 1, limit, Sort.by(Sort.Direction.DESC, "createdAt")));
        return PageResponse.of(pageResult.getContent(), page, limit, pageResult.getTotalElements());
    }

    public PageResponse<Article> findPublic(QueryArticleRequest query) {
        query.setStatus(ArticleStatus.published);
        return findAll(query);
    }

    public Article findOne(Long id) {
        return articleRepo.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Bài viết không tồn tại"));
    }

    public Article findBySlug(String slug) {
        return articleRepo.findBySlugAndStatus(slug, ArticleStatus.published)
                .orElseThrow(() -> new EntityNotFoundException("Bài viết không tồn tại"));
    }

    @Transactional
    public Article update(Long id, UpdateArticleRequest dto) {
        Article article = findOne(id);

        if (dto.getTitle() != null) article.setTitle(dto.getTitle());
        if (dto.getSlug() != null) article.setSlug(dto.getSlug());
        if (dto.getThumbnail() != null) article.setThumbnail(dto.getThumbnail());
        if (dto.getContent() != null) article.setContent(dto.getContent());
        if (dto.getExcerpt() != null) article.setExcerpt(dto.getExcerpt());
        if (dto.getStatus() != null) {
            article.setStatus(dto.getStatus());
            if (dto.getStatus() == ArticleStatus.published && article.getPublishedAt() == null) {
                article.setPublishedAt(LocalDateTime.now());
            }
        }
        if (dto.getSeoTitle() != null) article.setSeoTitle(dto.getSeoTitle());
        if (dto.getSeoDescription() != null) article.setSeoDescription(dto.getSeoDescription());
        if (dto.getCategoryId() != null) article.setCategoryId(dto.getCategoryId());

        if (dto.getTags() != null) {
            article.setTags(dto.getTags().isEmpty() ? new ArrayList<>() : findOrCreateTags(dto.getTags()));
        }

        return articleRepo.save(article);
    }

    @Transactional
    public void remove(Long id) {
        findOne(id);
        articleRepo.deleteById(id);
    }

    // News Categories
    public NewsCategory createCategory(String name, String slug, Integer sortOrder) {
        return newsCatRepo.save(NewsCategory.builder()
                .name(name).slug(slug).sortOrder(sortOrder != null ? sortOrder : 0).build());
    }

    public List<NewsCategory> findCategories() {
        return newsCatRepo.findAllByOrderBySortOrderAsc();
    }

    @Transactional
    public NewsCategory updateCategory(Long id, String name, String slug) {
        NewsCategory cat = newsCatRepo.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Danh mục tin tức không tồn tại"));
        if (name != null) cat.setName(name);
        if (slug != null) cat.setSlug(slug);
        return newsCatRepo.save(cat);
    }

    public void removeCategory(Long id) {
        newsCatRepo.deleteById(id);
    }

    private List<Tag> findOrCreateTags(List<String> names) {
        List<Tag> tags = new ArrayList<>();
        for (String name : names) {
            String slug = name.toLowerCase().replaceAll("[^a-z0-9\\p{L}]+", "-").replaceAll("^-|-$", "");
            Tag tag = tagRepo.findBySlug(slug).orElseGet(() ->
                    tagRepo.save(Tag.builder().name(name).slug(slug).build()));
            tags.add(tag);
        }
        return tags;
    }

    private Specification<Article> buildSpec(QueryArticleRequest query) {
        return (root, cq, cb) -> {
            List<Predicate> predicates = new ArrayList<>();
            if (query.getSearch() != null) {
                String like = "%" + query.getSearch() + "%";
                predicates.add(cb.or(cb.like(root.get("title"), like), cb.like(root.get("excerpt"), like)));
            }
            if (query.getCategoryId() != null) {
                predicates.add(cb.equal(root.get("categoryId"), query.getCategoryId()));
            }
            if (query.getTag() != null) {
                Join<Article, Tag> tagJoin = root.join("tags");
                predicates.add(cb.equal(tagJoin.get("slug"), query.getTag()));
            }
            if (query.getStatus() != null) {
                predicates.add(cb.equal(root.get("status"), query.getStatus()));
            }
            return cb.and(predicates.toArray(new Predicate[0]));
        };
    }
}
